stdlib/public/SwiftRemoteMirror/SwiftRemoteMirror.cpp

```diff
@@ -36,28 +36,37 @@ using namespace swift;
 using namespace swift::reflection;
 using namespace swift::remote;
 
-#if SWIFT_OBJC_INTEROP
-using Runtime = External<WithObjCInterop<RuntimeTarget<sizeof(uintptr_t)>>>;
-#else
-using Runtime = External<NoObjCInterop<RuntimeTarget<sizeof(uintptr_t)>>>;
-#endif
-using NativeReflectionContext = swift::reflection::ReflectionContext<Runtime>;
+using RuntimeWithObjCInterop =
+    External<WithObjCInterop<RuntimeTarget<sizeof(uintptr_t)>>>;
+using RuntimeNoObjCInterop =
+    External<NoObjCInterop<RuntimeTarget<sizeof(uintptr_t)>>>;
+
+using ReflectionContextWithObjCInterop =
+    swift::reflection::ReflectionContext<RuntimeWithObjCInterop>;
+using ReflectionContextNoObjCInterop =
+    swift::reflection::ReflectionContext<RuntimeNoObjCInterop>;
 
 struct SwiftReflectionContext {
-  NativeReflectionContext *nativeContext;
+  using ContextVariant = std::variant<ReflectionContextWithObjCInterop,
+                                      ReflectionContextNoObjCInterop>;
+
+  ContextVariant context;
   std::vector<std::function<void()>> freeFuncs;
   std::vector<std::tuple<swift_addr_t, swift_addr_t>> dataSegments;
 
   std::function<void(void)> freeTemporaryAllocation = [] {};
 
-  SwiftReflectionContext(MemoryReaderImpl impl) {
-    auto Reader = std::make_shared<CMemoryReader>(impl);
-    nativeContext = new NativeReflectionContext(Reader);
-  }
+  SwiftReflectionContext(bool objCInteropIsEnabled, MemoryReaderImpl impl)
+      : context(objCInteropIsEnabled
+                    ? ContextVariant(
+                          std::in_place_type<ReflectionContextWithObjCInterop>,
```

Comment:
This might be nicer looking if you skip the member initialization and use `if(objCInteropIsEnabled) context.emplace(...); else context.emplace(...);`. If that works.

Reply:
Right, it wants to construct it with `()` if you don't specify. You could make it work by having the variant store pointers (or `unique_ptr`) instead of the raw types, but that cure may be worse than the disease.

Reply:
Nice, that's actually sort of readable!